nat-lab/tests/test_batching.py

```diff
@@ -0,0 +1,281 @@
+        # peers will form a meshnet almost immediately, already achieving the desired batching effect.
+        for client in env.clients:
+            await client.stop_device()
+
+        tasks = []
+        for name in container_names:
+            task = asyncio.create_task(
+                capture_histograms(
+                    name,
+                    capture_duration,
+                    [
+                        PacketFilter("icmp", lambda p: p.haslayer(ICMP)),
+                        PacketFilter(
+                            "udp", lambda p: p.haslayer(IP) and p.haslayer(UDP)
+                        ),
+                        PacketFilter(
+                            "non-derp udp",
+                            lambda p: p.haslayer(IP)
+                            and p.haslayer(UDP)
+                            and not (p[UDP].sport == 3479 or p[UDP].dport == 3479),
+                        ),
+                    ],
+                )
+            )
+            tasks.append(task)
+
+        api = env.api
+
+        # misalign the peers by sleeping some before starting each node again
+        for client, node, _ in cnodes:
+            await asyncio.sleep(3)
+            await client.simple_start()
+            await client.set_meshmap(api.get_meshmap(node.id))
+
+        await asyncio.gather(
+            alpha_client.wait_for_state_peer(
+                beta.public_key,
+                [State.Connected],
+                [PathType.Direct],
+            ),
+            alpha_client.wait_for_state_peer(
+                gamma.public_key,
+                [State.Connected],
+                [PathType.Direct],
+            ),
+            beta_client.wait_for_state_peer(
+                alpha.public_key,
+                [State.Connected],
+                [PathType.Direct],
+            ),
+            beta_client.wait_for_state_peer(
+                gamma.public_key,
+                [State.Connected],
+                [PathType.Direct],
+            ),
+            gamma_client.wait_for_state_peer(
+                alpha.public_key,
+                [State.Connected],
+                [PathType.Direct],
+            ),
+            gamma_client.wait_for_state_peer(
+                beta.public_key,
+                [State.Connected],
+                [PathType.Direct],
+            ),
+        )
+
+        futures = await asyncio.gather(*tasks)
+        for name, histograms in zip(container_names, futures):
+            for filter_name, hs in histograms:
+                title = f"{name}-filter({filter_name})-id({request.node.callspec.id})"
+                draw_histogram(title, hs, max_height=12)
+                silence_streak = longest_zero_streak(hs)
+                print(f"Longest silence for {title} is {silence_streak}\n")
+
+        assert (
+            False
+        ), "TODO: implement a proper assert(this assert also forced the dump of output)"
```

Comment:
So far the results are not yet as we expect. There's Pyro5 traffic, there's STUN traffic from the DERP server, etc. For now the assertion is always False in order to dump the test output and in turn observe the histograms rendered in ascii

Reply:
An alternative might be to save the drawings alongside artifacts but I think ascii provides a smooth way of doing the same at the cost of `-v` flag on `run_local.py`